### PR TITLE
[14.0][FIX] sale_purchase_force_vendor: Display the vendor_id field only 1 time so that it is always displayed and apply the correct domain in both use cases

### PR DESCRIPTION
**Pull request by victoralmau**
Display the `vendor_id` field only 1 time so that it is always displayed and apply the correct domain in both use cases

Please @pedrobaeza and @chienandalu can you review it?

@Tecnativa TT39679

### CODE REVIEW

**Comment by victoralmau** [comment]
Ping @chienandalu

File: sale_purchase_force_vendor/models/sale_order_line.py [hunk 1]
```diff
@@ -27,15 +27,23 @@ class SaleOrderLine(models.Model):
     @api.depends("product_id")
     def _compute_allowed_vendor_ids(self):
         supplierinfo_model = self.env["product.supplierinfo"]
+        partner_model = self.env["res.partner"]
         self.allowed_vendor_ids = self.allowed_vendor_ids
-        for item in self.filtered(lambda x: x.product_id):
+        for item in self.filtered(
+            lambda x: x.product_id and x.order_id.sale_purchase_force_vendor_restrict
```
Comment (pedrobaeza) [inline sale_purchase_force_vendor/models/sale_order_line.py]: If you are here filtering section and notes, you should filter the proper field (`display_type`?)

File: sale_purchase_force_vendor/models/sale_order_line.py [hunk 2]
```diff
@@ -27,15 +27,23 @@ class SaleOrderLine(models.Model):
...
+        ):
             item.allowed_vendor_ids = supplierinfo_model.search(
                 [
                     "|",
                     ("product_tmpl_id", "=", item.product_id.product_tmpl_id.id),
                     ("product_id", "=", item.product_id.id),
                 ]
             ).mapped("name")
+        for item in self.filtered(
```
Comment (pedrobaeza) [inline sale_purchase_force_vendor/models/sale_order_line.py]: Please don't do two times the same `filtered` for performance reason.

File: sale_purchase_force_vendor/models/sale_order_line.py [hunk 3]
```diff
@@ -27,15 +27,23 @@ class SaleOrderLine(models.Model):
...
+            lambda x: x.product_id
+            and not x.order_id.sale_purchase_force_vendor_restrict
+        ):
+            item.allowed_vendor_ids = partner_model.search([])
```
Comment (pedrobaeza) [inline sale_purchase_force_vendor/models/sale_order_line.py]: This is totally overkill on a populated database. Isn't there other solution?

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by victoralmau** [comment]
I have changed the approach, what do you think now?

**Comment by victoralmau** [comment]
Ping @pedrobaeza

File: sale_purchase_force_vendor/models/sale_order_line.py [hunk 4]
```diff
@@ -18,17 +20,36 @@ class SaleOrderLine(models.Model):
...
+        for item in self:
+            domain = (
+                [("id", "in", item.allowed_vendor_ids.ids)]
+                if item.order_id.sale_purchase_force_vendor_restrict
+                else [(1, "=", 1)]
+            )
+            item.vendor_id_domain = json.dumps(domain)
+
     @api.depends("product_id")
     def _compute_allowed_vendor_ids(self):
```
Comment (pedrobaeza) [inline sale_purchase_force_vendor/models/sale_order_line.py]: Why keeping this field?

Comment (victoralmau) [reply sale_purchase_force_vendor/models/sale_order_line.py]: You are right, `allowed_vendor_ids` field is no longer needed, it is already changed.

File: sale_purchase_force_vendor/views/sale_order_view.xml [hunk 5]
```diff
@@ -12,33 +12,22 @@
...
-                    attrs="{'column_invisible': [('parent.sale_purchase_force_vendor_restrict', '=', False)]}"
-                    domain="[('id','in', allowed_vendor_ids)]"
-                    options='{"no_create": True}'
-                />
-                <field
-                    name="vendor_id"
-                    attrs="{'column_invisible': [('parent.sale_purchase_force_vendor_restrict', '=', True)]}"
+                    domain="vendor_id_domain"
                     options='{"no_create": True}'
+                    optional="hide"
```
Comment (pedrobaeza) [inline sale_purchase_force_vendor/views/sale_order_view.xml:20]: This one should be invisible.

File: sale_purchase_force_vendor/tests/common.py [hunk 6]
```diff
@@ -8,6 +8,7 @@ class TestSalePurchaseForceVendorBase(common.SavepointCase):
     @classmethod
     def setUpClass(cls):
         super().setUpClass()
+        cls.company = cls.env.company
```
Comment (pedrobaeza) [inline sale_purchase_force_vendor/tests/common.py]: Do you really need it?

File: sale_purchase_force_vendor/models/sale_order_line.py [hunk 7]
```diff
@@ -18,24 +20,21 @@ class SaleOrderLine(models.Model):
...
-                ]
-            ).mapped("name")
+    def _compute_vendor_id_domain(self):
+        for item in self:
+            domain = (
+                [("id", "in", item.product_id.variant_seller_ids.name.ids)]
+                if item.order_id.sale_purchase_force_vendor_restrict
+                else [(1, "=", 1)]
+            )
+            item.vendor_id_domain = json.dumps(domain)
```
Comment (pedrobaeza) [inline sale_purchase_force_vendor/models/sale_order_line.py:37]: Is json dump needed?

File: sale_purchase_force_vendor/models/sale_order_line.py [hunk 8]
```diff
@@ -18,24 +20,21 @@ class SaleOrderLine(models.Model):
...
-                    ("product_tmpl_id", "=", item.product_id.product_tmpl_id.id),
-                    ("product_id", "=", item.product_id.id),
-                ]
-            ).mapped("name")
+    def _compute_vendor_id_domain(self):
+        for item in self:
+            domain = (
+                [("id", "in", item.product_id.variant_seller_ids.name.ids)]
+                if item.order_id.sale_purchase_force_vendor_restrict
+                else [(1, "=", 1)]
```
Comment (pedrobaeza) [inline sale_purchase_force_vendor/models/sale_order_line.py]: Isn't `[]` working for empty domain and thus, searching for everything?

File: sale_purchase_force_vendor/views/sale_order_view.xml (hunk 5 above)
Comment (victoralmau) [reply sale_purchase_force_vendor/views/sale_order_view.xml:20]: According to previous approach `vendor_id` field to be displayed.

File: sale_purchase_force_vendor/models/sale_order_line.py (hunk 7 above)
Comment (victoralmau) [reply sale_purchase_force_vendor/models/sale_order_line.py:37]: Yes, according to documentation: https://github.com/OCA/web/tree/14.0/web_domain_field

File: sale_purchase_force_vendor/views/sale_order_view.xml (hunk 5 above)
Comment (pedrobaeza) [reply sale_purchase_force_vendor/views/sale_order_view.xml:20]: Sorry, I have mixed the field `vendor_id_domain`.

**Review by pedrobaeza — APPROVED** [review]

**Review by chienandalu — APPROVED** [review]
Nice :+1:

**Pull request merged**